kernel-rs/src/vm.rs

```diff
@@ -21,18 +23,437 @@ extern "C" {
+            if n > max {
+                n = max
+            }
+            let mut p: *mut u8 = pa0.wrapping_add(srcva.wrapping_sub(va0)) as *mut u8;
+            while n > 0 {
+                if *p as i32 == '\u{0}' as i32 {
+                    *dst = '\u{0}' as i32 as u8;
+                    got_null = 1;
+                    break;
+                } else {
+                    *dst = *p;
+                    n = n.wrapping_sub(1);
+                    max = max.wrapping_sub(1);
+                    p = p.offset(1);
+                    dst = dst.offset(1)
+                }
+            }
+            srcva = va0.wrapping_add(PGSIZE)
+        }
+        if got_null != 0 {
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+}
+
+pub struct PageTable {
+    ptr: *mut RawPageTable,
+}
+
+impl PageTable {
+    pub fn new() -> Self {
+        let page = unsafe { kalloc() } as *mut RawPageTable;
+        if page.is_null() {
+            panic!("PageTable new: out of memory");
+        }
+        unsafe {
+            ptr::write_bytes(page, 0, 1);
+        }
+
+        Self { ptr: page }
+    }
+
+    pub fn from_raw(ptr: *mut RawPageTable) -> Self {
+        Self { ptr }
+    }
+
+    pub fn into_raw(self) -> *mut RawPageTable {
+        let ret = self.ptr;
+        mem::forget(self);
+        ret
+    }
+
+    pub fn is_null(&self) -> bool {
+        self.ptr.is_null()
+    }
+
+    pub fn as_raw(&self) -> *mut RawPageTable {
+        self.ptr
+    }
+}
+
+// TODO: separate these function for va structure later (Use type to show this)
+impl PageTable {
+    /// Create PTEs for virtual addresses starting at va that refer to
+    /// physical addresses starting at pa. va and size might not
+    /// be page-aligned. Returns Ok(()) on success, Err(()) if walk() couldn't
+    /// allocate a needed page-table page.
+    pub unsafe fn mappages(
+        &mut self,
+        va: usize,
+        size: usize,
+        mut pa: usize,
+        perm: i32,
+    ) -> Result<(), ()> {
+        let mut a = pgrounddown(va);
+        let last = pgrounddown(va.wrapping_add(size).wrapping_sub(1usize));
+        loop {
+            let pte = some_or!(walk(self, a, 1), return Err(()));
+            if pte.check_flag(PTE_V) {
+                panic!("remap");
+            }
+            pte.set_inner(pa2pte(pa) | perm as usize | PTE_V);
+            if a == last {
+                break;
+            }
+            a = a.wrapping_add(PGSIZE);
+            pa = pa.wrapping_add(PGSIZE);
+        }
+        Ok(())
+    }
+
+    /// Load the user initcode into address 0 of pagetable,
+    /// for the very first process.
+    /// sz must be less than a page.
+    pub unsafe fn uvminit(&mut self, src: *mut u8, sz: u32) {
+        if sz >= PGSIZE as u32 {
+            panic!("inituvm: more than a page");
+        }
+        let mem: *mut u8 = kalloc() as *mut u8;
+        ptr::write_bytes(mem as *mut libc::CVoid, 0, PGSIZE);
```

Comment:
```suggestion
        ptr::write_bytes(mem, 0, PGSIZE);
```
`u8` 과 `CVoid` 모두 동일한 크기 (1바이트) 입니다.